File: android/src/main/java/com/websquare/plugins/zendesksupport/ZendeskSupportPlugin.kt
package com.websquare.plugins.zendesksupport

import com.getcapacitor.Plugin
import com.getcapacitor.annotation.CapacitorPlugin
import com.websquare.plugins.zendesksupport.ZendeskSupport
import com.getcapacitor.PluginMethod
import com.getcapacitor.PluginCall
import org.json.JSONArray
import org.json.JSONException
import java.lang.Exception
import java.util.ArrayList

@CapacitorPlugin(name = "ZendeskSupport")
class ZendeskSupportPlugin : Plugin() {
    private val implementation = ZendeskSupport()
    @PluginMethod
    fun initialize(call: PluginCall) {
        val appId = call.getString("appId", "")
        val clientId = call.getString("clientId", "")
        val zendeskUrl = call.getString("zendeskUrl", "")
        val debugLog = call.getBoolean("debugLog", false)
        try {
            implementation.initialize(activity.applicationContext, zendeskUrl, appId, clientId, debugLog)
            call.resolve()
        } catch (e: Exception) {
            call.reject(e.message, e)
        }
    }

    @PluginMethod
    fun setAnonymousIdentity(call: PluginCall) {
        val name = call.getString("name", "")
        val email = call.getString("email", "")
        implementation.setAnonymousIdentity(name, email)
        call.resolve()
    }

    @PluginMethod
    fun setIdentity(call: PluginCall) {
        val token = call.getString("token", "")
        implementation.setIdentity(token)
        call.resolve()
    }

    @PluginMethod
    fun showHelpCenter(call: PluginCall) {
        val groupBy = call.getString("groupBy", "")
        var groupIds: List<Long?> = ArrayList()
        var labels: List<String?> = ArrayList()
        if (call.hasOption("groupIds")) {
            groupIds = jsonArrayToList(call.getArray("groupIds"))
        }
        if (call.hasOption("labels")) {
            labels = jsonArrayToList(call.getArray("labels"))
        }
        implementation.showHelpCenter(activity, groupBy, groupIds, labels)
        call.resolve()
    }

    @PluginMethod
    fun showHelpCenterArticle(call: PluginCall) {
        val articleId = call.getString("articleId", "")
        implementation.showHelpCenterArticle(activity, articleId)
        call.resolve()
    }

    @PluginMethod
    fun showTicketRequest(call: PluginCall) {
        val subject = call.getString("subject", "")
        var tags: List<String?> = ArrayList()
        var fields: List<String?> = ArrayList()
        if (call.hasOption("tags")) {
            tags = jsonArrayToList(call.getArray("tags"))
        }
        if (call.hasOption("fields")) {
            fields = jsonArrayToList(call.getArray("fields"))
        }
        implementation.showTicketRequest(activity, subject, tags, fields)
        call.resolve()
    }

    @PluginMethod
    fun showUserTickets(call: PluginCall) {
        implementation.showUserTickets(activity)
        call.resolve()
    }

    private fun <T> jsonArrayToList(jsonArray: JSONArray): List<T> {
        val arrayList: MutableList<T> = ArrayList()
        for (i in 0 until jsonArray.length()) {
            try {
                arrayList.add(jsonArray[i] as T)
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }
        return arrayList
    }
}

File: ios/Plugin/ZendeskSupportPlugin.swift
import Foundation
import Capacitor
import SupportSDK
import ZendeskCoreSDK
import UIKit
/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(ZendeskSupportPlugin)
public class ZendeskSupportPlugin: CAPPlugin, UINavigationControllerDelegate {
    private let implementation = ZendeskSupport()

    @objc func initialize(_ call: CAPPluginCall) {
        do {
            let appId = call.getString("appId") ?? ""
            let clientId = call.getString("clientId") ?? ""
            let zendeskUrl = call.getString("zendeskUrl") ?? ""
            let debugLog = call.getBool("debugLog") ?? false
            try implementation.initialize( appId, clientId, zendeskUrl, debugLog)
            call.resolve()
        } catch {
            call.reject(error.localizedDescription, nil, error)
        }
    }
    
    @objc func setAnonymousIdentity(_ call: CAPPluginCall) {
        let name = call.getString("name") ?? ""
        let email = call.getString("email") ?? ""
        implementation.setAnonymousIdentity(name, email)
        call.resolve()
    }
    
    @objc func setIdentity(_ call: CAPPluginCall) {
        let token = call.getString("token") ?? ""
        implementation.setIdentity(token)
        call.resolve()
    }
    
    @objc func showHelpCenter(_ call: CAPPluginCall) {
        let groupBy = call.getString("groupBy") ?? ""
        let ids = call.getArray("groupIds", [])
        let lbls = call.getArray("labels", [])
        
        var groupIds: [NSNumber] = []
        var labels: [String] = []
        guard let bridge = self.bridge else {
            call.resolve()
            return
        }
        
        if (!ids.isEmpty) {
            for id in ids {
                if let gId = id as? NSNumber {
                    groupIds.append(gId)
                }
            }
        }
        if (!lbls.isEmpty) {
            for lbl in lbls {
                if let label = lbl as? String {
                    labels.append(label)
                }
            }
        }
        implementation.showHelpCenter(bridge.viewController, groupBy, groupIds, labels)
        call.resolve()
    }
    
    @objc func showHelpCenterArticle(_ call: CAPPluginCall) {
        let articleId = call.getString("articleId") ?? ""
        implementation.showHelpCenterArticle(bridge?.viewController, articleId)
        call.resolve()
    }
    
    @objc func showTicketRequest(_ call: CAPPluginCall) {
        let subject = call.getString("subject") ?? ""
        let tgs = call.getArray("tags", [])
        let flds = call.getArray("fields", [])
        
        var tags: [String] = []
        var fields: [String] = []
        if (!tgs.isEmpty) {
            for tg in tgs {
                if let tag = tg as? String {
                    tags.append(tag)
                }
            }
        }
        
        if (!flds.isEmpty) {
            for fld in flds {
                if let field = fld as? String {
                    fields.append(field)
                }
            }
        }
        
        implementation.showTicketRequest(bridge?.viewController, subject, tags, fields)
    }
    
    @objc func showUserTickets(_ call: CAPPluginCall) {
        implementation.showUserTickets(bridge?.viewController)
    }
}
